test: Add WithContext helper

aanm opened the pull request:
This PR introduces `context.Context` into the testing framework. It will help debugging https://github.com/cilium/cilium/issues/5812 once this PR is merged and will help f.i.x. https://github.com/cilium/cilium/issues/5163

The changes are best viewed as a whole, instead a per-commit basis

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5821)
<!-- Reviewable:end -->

aanm committed:
test: add WithContext helper

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: make use of RunCommandContext

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: add WaitGroup delta before goroutine

To avoid a WaitGroup.Wait() being executed without the
execution of Add(1), the Add(1) should be perform before a go routine
initialization.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: refactor ExecContext to ExecInBackground

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: add sshMeta.ExecContext to execute commands with context

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: add CiliumExecContext

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: make CiliumEndpointsList use context.Context

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: add error field to CmdRes

With this commit it will allow to store the error of a command execution.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: make use of Context in CiliumEndpointWaitReady

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on test/helpers/cons.go:
```diff
@@ -178,7 +178,7 @@ const (
 
```
Remove this commit.

aanm replied:
Done

ianvernon commented on test/helpers/cmd.go:
```diff
@@ -190,7 +190,12 @@ func (res *CmdRes) Filter(filter string) (*FilterBuffer, error) {
 
 // ByLines returns res's stdout split by the newline character .
```
Update documentation to include the return character as well.

ianvernon commented on test/helpers/cons.go:
```diff
@@ -28,6 +29,8 @@ import (
 	// HelperTimeout is a predefined timeout value for commands, in seconds.
 	HelperTimeout int64 = 300
+	// HelperTimeoutDuration is a predefined timeout value for commands.
+	HelperTimeoutDuration = time.Duration(HelperTimeout) * time.Second
```
HelperTimeout was initially a duration, but we moved away from that due to 460c376e45813f77285ea2cc6c23c7854a30807c - I'd prefer to keep one or the other, otherwise I feel like we will run into the same issues that that commit addressed.

aanm replied:
I disagree, this clearly states this is a `time.Duration`. Commit 460c376 fixed a different issue related with semantics where we had a Duration without the definition of the time unit set. One of the fixes could have been the one introduced in 460c376, the other could have been the change the of `TimeoutConfig`. As we are moving towards `context.Context` the `HelperTimeout` will likely be deleted.

ianvernon commented on test/helpers/cmd.go:
```diff
@@ -42,6 +42,7 @@ type CmdRes struct {
 	exitcode int             // The exit code of cmd
```
Type in commit: "the error of a command execution"

aanm committed:
test: Refactor RunCommandContext to RunCommandContextCancel

This will allow to use RunCommandContextCancel for background / async
operations and RunCommandContext for context executions.
RunCommandContextCancel can be canceled without returning the error of
the context.Context function, if RunCommandContext's context is
canceled, RunCommandContext will return the "Context" canceled error.

Signed-off-by: André Martins <andre@cilium.io>

vadorovsky commented on test/helpers/cmd.go:
```diff
@@ -140,7 +141,7 @@ func (res *CmdRes) CombineOutput() *bytes.Buffer {
 // IntOutput returns the stdout of res as an integer
 func (res *CmdRes) IntOutput() (int, error) {
-	return strconv.Atoi(strings.Trim(res.stdout.String(), "\n"))
+	return strconv.Atoi(strings.Trim(strings.Trim(res.stdout.String(), "\n"), "\r"))
```
AFAIK `strings.Trim` trims all characters from the string provided in the second argument, so I think that one call should be enough:

```
return trconv.Atoi(strings.Trim(res.stdout.String(), "\n\r"))
```

vadorovsky commented on test/helpers/node.go:
```diff
@@ -116,10 +116,11 @@ func (s *SSHMeta) setBasePath() {
-// the command in the provided buffers. Returns false if the command failed
-// during its execution.
-func (s *SSHMeta) Execute(cmd string, stdout io.Writer, stderr io.Writer) error {
+// ExecuteContext executes the given `cmd` and writes the cmd's stdout and sterr
```
Typo:
sterr -> stderr

vadorovsky reviewed (changes requested):
minor nits
